salt/utils/data.py

```diff
@@ -767,15 +765,19 @@ def traverse_dict_and_list(data, key, default=None, delimiter=DEFAULT_TARGET_DEL
                     ptr = ptr[idx]
                 except IndexError:
                     return default
-        else:
+        elif isinstance(ptr, dict):
             try:
-                if type(ptr) in (datetime, datetime.date, datetime.datetime, datetime.time):
-                    ptr = salt.utils.json.loads(salt.utils.json.dumps(ptr, indent=4, sort_keys=True, default=str))
-                else:
-                    ptr = salt.utils.json.loads(salt.utils.json.dumps(ptr))
+                for key, value in ptr.items():
+                    if not isinstance(key,str):
+                        ptr[str(key)] = ptr.pop(key)
```

Comment:
I am not 100% sure that this code is doing the same thing that the one removed, especially since it removes the items of type `str` which is not even the same thing in python 2 and 3.

In any case this PR needs unit tests showing the problem without the fix and showing it fixed.

Reply:
while resolving the test that fails, I face errors like DateTime object is not serializable that's why I change the code from ```JSON dump``` to rechanging the dict key type to the ```string``` and assigned it with the poped value of the original key. The main thing in the issue is that the key which is of type other than ```str``` is not accessible. Because in ```pillar.get``` they pass a string with separator as an argument like this 'redis_advanced2:6379:maxmemory'. so here the integer type key is not accessible. But it occurs in pillar - The key of type integer 6379 ```redis_advanced:
  '6379':
    max memory: 2147483648
redis_advanced2:
  6379:
    max memory: 2147483648```
So I thought to convert the key type to string will resolve the mentioned issue. I tested the function traverse_dict_and_list which calls from the pillar. get function. I verify the output too. The code change works in both python 2 and python3. please check [here](https://docs.google.com/document/d/1dbes9TvXtmd0aDyWIP2voTjCyIeyK1L96s_zSvzhAYk/edit?usp=sharing), I tested against the input mentioned in issue #56444.